src/ripple/app/ledger/impl/ApplyView.cpp

```diff
@@ -0,0 +1,296 @@
+
+    Permission to use, copy, modify, and/or distribute this software for any
+    purpose  with  or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#include <BeastConfig.h>
+#include <ripple/ledger/ApplyView.h>
+#include <ripple/basics/contract.h>
+#include <ripple/protocol/Protocol.h>
+#include <cassert>
+
+namespace ripple {
+
+boost::optional<std::uint64_t>
+ApplyView::dirInsert (
+    Keylet const& directory,
+    uint256 const& key,
+    bool strictOrder,
+    std::function<void (std::shared_ptr<SLE> const&)> describe)
+{
+    auto root = peek(directory);
+
+    if (! root)
+    {
+        // No root, make it.
+        root = std::make_shared<SLE>(directory);
+        root->setFieldH256 (sfRootIndex, directory.key);
+        describe (root);
+
+        STVector256 v;
+        v.push_back (key);
+        root->setFieldV256 (sfIndexes, v);
+
+        insert (root);
+        return std::uint64_t{0};
+    }
+
+    std::uint64_t page = root->getFieldU64(sfIndexPrevious);
+
+    auto node = root;
+
+    if (page)
+    {
+        node = peek (keylet::page(directory, page));
+        if (!node)
+            LogicError ("Directory chain: root back-pointer broken.");
+    }
+
+    auto indexes = node->getFieldV256(sfIndexes);
+
+    // If there's space, we use it:
+    if (indexes.size () < dirNodeMaxEntries)
+    {
+        if (!strictOrder)
+        {
+            // We can't be sure if this page is already sorted because
+            // it may be a legacy page we haven't yet touched. Take
+            // the time to sort it.
+            std::sort (indexes.begin(), indexes.end());
+
+            auto pos = std::lower_bound(indexes.begin(), indexes.end(), key);
+
+            if (pos != indexes.end() && key == *pos)
+                LogicError ("dirInsert: double insertion");
+
+            indexes.insert (pos, key);
+        }
+        else
+        {
+            if (std::find(indexes.begin(), indexes.end(), key) != indexes.end())
+                LogicError ("dirInsert: double insertion");
+
+            indexes.push_back(key);
+        }
+
+        node->setFieldV256 (sfIndexes, indexes);
+        update(node);
+
+        return page;
+    }
+
+    // Check whether we're out of pages.
+    if (++page == dirNodeMaxPages)
```

Comment:
No, since pages are zero-based. But I guess making it `>= dirNodeMaxPages` is a good idea anyways. There's no directory with that many pages, but using `>=` will do the right thing if there were..